Applications/Spire/Include/Spire/Ui/TextBox.hpp

```diff
@@ -0,0 +1,68 @@
+#ifndef SPIRE_TEXT_BOX_HPP
+#define SPIRE_TEXT_BOX_HPP
+#include <QLineEdit>
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  //! Displays a one-line text box.
+  class TextBox : public QLineEdit {
+    public:
+
+      //! Signals that the current text changed.
+      using CurrentSignal = Signal<void (const QString& text)>;
```

Comment:
Good question. I thought about this question and tried to do it. The current signal can be replaced by the QLineEdit::textEdited signal, but I can’t find a way to replace the submit signal. Actually, there is a QLineEdit::editingFinished signal which is similar to the submit signal. The reason that I can’t use it directly is that the signal can be emitted when the textbox is read-only. I suspect it’s a Qt’s bug. As a result, I have to filter the signal manually in this particular case. For consistency, I chose the Spire signal for these two signals.  Do you have any good ideas for the editingFinished signal?

Reply:
If it makes the implementation unnecessarily complicated it's probably fine to leave it as it is.

Reply:
Why do we have to use boost signals? We can define Qt signals in the class.

Reply:
I don't think we use Qt's meta-object compiler when building Spire, but @kranar will know for sure.

Reply:
We use Qt because it's basically the only option for cross platform UIs in C++, not because it's particularly good. When there is an option to avoid using Qt to implement a feature, such as the standard library or another third party library, we prefer to use it.